### PR TITLE
ERV-43 - Update code to ES6 (WIP)

### PR DESCRIPTION
**Pull request by turisap**
Updates:

- create class `Helpers` and put all functions inside as static methods
- replaced `var` with `let` and `const`
- replaces anonymous callback functions with arrow functions
- replaced `require` with `import`
- replaced `module.exports` with `export default class`
- replaced `Object.assign()` in ` reverseLineString()` with spread operator

Need your feedback on whether on not I'm heading the right direction :-)

Can I add `selenium-debug.log` to `.gitignore`?

### CODE REVIEW

**Comment by qooban** [comment]
>Need your feedback on whether on not I'm heading the right direction :-)
Yes, that's exactly what I was thinking about :-) 
1. What will we need in terms of builder - I'm currently using `browserify` and `uglify`? Not sure they're working fine with ES6.
2. Have you managed to run tests? (see description in readme)

>Can I add selenium-debug.log to .gitignore?
Yes, sure.

**Comment by turisap** [comment]
> What will we need in terms of builder - I'm currently using browserify and uglify? Not sure they're working fine with ES6.

I actually haven't been working with both of them, but use `wepback` and `babel` in my React.js apps, so I could try to deploy it here as well. I'm not completely sure though what is the goal.. Do you need to get a bundle file out of them all as if with running `npm run bundle:test`? If so, I guess `webpack` and `babel` can do the job.

> Have you managed to run tests? (see description in readme)

Yes, I did on a freshly cloned repo, but now they are failing. I think this is due to I need to revise other files as well and make some changes in import as well.

So, I'll proceed to update to ES6 the other files for now.

**Comment by qooban** [comment]
> I'm not completely sure though what is the goal.. Do you need to get a bundle file out of them all as if with running npm run bundle:test? If so, I guess webpack and babel can do the job.

Yeah, I want bundle file to be the output. It is used by another application (`edk-software/rejony-app`). I'm also using webpack and babel in my other project, but decided to use browserify and uglifyjs here as I thought it would be easier. If you can do it (create bundle to be used from the browser + minimize it) with webpack and babel, let's give it a try.

> Yes, I did on a freshly cloned repo, but now they are failing.
Looks like the test page is not loaded properly. Do you have test server properly started (`npm start`) and can you access `http://localhost:7777` ?

**Comment by turisap** [comment]
> Looks like the test page is not loaded properly. Do you have test server properly started (npm start) and can you access http://localhost:7777 ?

Yes, I followed the instructions and run  `npm install`, then  `npm start` and what I get in my terminal is
```
> edk-route-verifier@1.0.1 start D:\IT\Code\edk-route-verifier
> cd server && npm run start


> proxy@1.0.0 start D:\IT\Code\edk-route-verifier\server
> node server.js -c ./config.json

Starting proxy server at: http://localhost:7777
Sending KML for route 1.
```

When I visited `localhost:7777` I got an error on the screen as "cannot GET /" and no errors in the console.

**Comment by qooban** [comment]
My bad, there's no `/` route defined. 

Please try:
* `localhost:7777/resources` - list all test resources available
* `localhost:7777/1` - shows page for `1.kml` test file

**Comment by turisap** [comment]
These two work well!

**Comment by turisap** [comment]
@qooban, I've updated this pull request with new commits. I've updated all files in `src` directory to ES6 syntax, please take a look. There are two possible problems, though:
- in `src/Route.js` I have moved checks `isEmpty()` and `isRouteVerifiable` to the class constructor and the last one just returns in the case of an unverifiable route. I guess it worked in the old function constructor call implementation, but I'm not sure how it can affect execution of the code now. (I put comments in the class's constructor for you to easier to find it)
- I run `npm test` now and it looks like changes in the code does not affect tests results, same error and, as I noticed now, there are some tests skipped, like:
```
 SKIPPED:
   - Positive test - circular route
   - Positive test - zero-leading station numbers
   - Positive test - long path
   - Negative test - duplicated station point
   - Negative test - station out of path
   - Positive test - circular path
   - Negative test - stations in reversed order comparing to path direction
   - Negative test - 20-duplicated path
   - Negative test - 21-two path
   - Negative test - 22-15_stations
   - Negative test - 23-13_stations
   - Positive test - 24-Short_distance_between_1_14
   - Positive test - 25-eight_shaped_route
```
does that mean that many of the tests simply haven't been run? And what should I do next apart of `webpack` implementation?

**Comment by qooban** [comment]
@turisap Thanks! I'll take a look in my free time.

> does that mean that many of the tests simply haven't been run? 

Yes. They haven't been run, because previous failure.

> And what should I do next apart of webpack implementation?

I think webpack implementation would be part of updating code to ES6 as tests must be passing before merging.

As you probably know - you need to have `edk-route-verifier.js.min` bundle in `server/static/js` directory. It's done by `npm run bundle:test` and it's not working now with ES6 code, right? When you got it working and still tests will be failing, please send me logs and let me know. 

BTW I'm assuming you have created your `config.json` with generated Google Maps API key and added necessary Google Maps libs, right?

**Comment by turisap** [comment]
@qooban, I'm happy to participate here :-)

> As you probably know - you need to have edk-route-verifier.js.min bundle in server/static/js directory. It's done by npm run bundle:test and it's not working now with ES6 code, right? 

Yes, that's right, if I run `npm run build:test` it trows a `ParseError` and does not build anything.

> BTW I'm assuming you have created your config.json with generated Google Maps API key and added necessary Google Maps libs, right?

Yes, `config.json` is in place with an API from Google and all three APIs (Embed, Elevation, JavaScript) enabled.

I've finished `webpack` implementation and got minified `edk-route-verifier.min.js` bundle in `server/static/js` directory. It's not clear for now how can I run tests on this new bundle. If I run just `npm test` it trows the same error as before:
```
 √ Element <body> was visible after 53 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 47 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 18 milliseconds.
 × Timed out while waiting for element <div#pageReloadModal> to be visible for 5000 milliseconds.  - expected "visible" but got: "not visible"
```

**Comment by qooban** [comment]
> Yes, that's right, if I run npm run build:test it trows a ParseError and does not build anything.

Actually I just checked and we expect `edk-route-verifier.js` bundle. This name can be configured in `server/server.js` - see part:
```
    res.render('pages/index', {
        googleMapsApiKey: configuration.googleMapsApiKey,
        routeId: id,
        serverPort: port,
        jsFilename: 'edk-route-verifier.js'
    });
```
Sorry for mistake.

I have just checked (created bundle using webpack) and still there is a problem. 

What I suggest is:
1. Update test server configuration (bundle name in `jsFilename`) or webpack configuration to work with each other
2. Create bundle
3. Start test server (`npm run start`)
4. Check errors at `http://localhost:7777/1` after load and when you click on the button:
obraz
Normally it should work - the site should look similarly to that:
obraz

I would suggest to first resolve that problem for single route, then try running all tests.

Reg. OS - Windows shall be OK. I am developing it on Windows and Linux. `npm install` should fetch selenium server standalone and chrome driver for you and nightwatch configuration I have written to be platform independent.

When the new version is stable and all tests pass, I'd like to ask you also to update `npm run build:*` to use webpack (eg. call new tasks you created) and main readme file.

**Comment by turisap** [comment]
Well, I've got good news: after some debugging all test pass now :-)

> When the new version is stable and all tests pass, I'd like to ask you also to update npm run build:* to use webpack (eg. call new tasks you created) and main readme file.

I can create two options, like `npm run build:dev` and `npm run build:prod` with different quality of source maps for example, or any other options you'd like to see. As another option, I can make just one)

**Comment by qooban** [comment]
Great! :)

>I can create two options, like npm run build:dev and npm run build:prod with different quality of source maps for example, or any other options you'd like to see. As another option, I can make just one)

I like your proposal - `npm run build:dev` and `npm run build:prod`.

**Comment by qooban** [comment]
In case of production version the size optimization is the most important.

**Comment by turisap** [comment]
@qooban, I have finished updates and pushed all the changes, please take a look.

As you can see in `package.json`, there are no two extra scripts: `build:prod` and `build:dev`. The former creates a bundle of 86 kB and another file with source maps around 800 kB, the latter creates one single file 1.85Mb with source maps inside. Is  86 kB what you expected?

Please review and let me know if there is something wrong :-)

**Comment by qooban** [comment]
@turisap Great! Thanks a lot! You're doing great job. I'll take a look. Cannot promise that I'll do that during weekend, but on Monday/Tuesday I shall have some time for that.

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,8 @@ node_modules/
 /test/screenshots/
 /test/reports/
 
+# Debug logs
+selenium-debug.log
+
+# Keys
+server/config.json
```
Comment (qooban) [inline .gitignore]: Isn't it already included few lines above?

File: package.json [hunk 2]
```diff
@@ -23,21 +25,28 @@
     "url": "https://github.com/edk-software/edk-route-verifier.git"
   },
   "dependencies": {
+    "@babel/core": "^7.0.0-beta.40",
+    "@babel/preset-env": "^7.0.0-beta.40",
```
Comment (qooban) [inline package.json]: I'm wondering shall we use beta version here?
Let's try with latest stable on all dependencies and see how it works.

BTW Thanks for sorting deps 👍

File: src/Route.js [hunk 3]
```diff
@@ -1,92 +1,104 @@
...
+        this.lineString = helpers.getLineString(this.geoJson);
+        this.points = helpers.getPoints(this.geoJson);
+        this.isRouteVerifiable = true;
+        this.numberOfPaths = helpers.getNumberOfFeatures('LineString', this.geoJson);
+        this.stations = new Stations(this.points, this.lineString);
+        this.path = this.stations.isPathReversed() ? helpers.reverseLineString(this.lineString) : this.lineString;
+        this.pathElevation;
+
+        /**
+         * I'm not sure if this is the right way to implement these checks
```
Comment (qooban) [inline src/Route.js]: We must think it over. These checks were to avoid initializing `stations` and `path` properties when there is no linestring or points in KML file. So the order in previous version matters. Try to upload KML without linestring or without points. How it will behave?

File: src/Stations.js [hunk 4]
```diff
@@ -1,69 +1,91 @@
...
             }
 
             // split
-            var parts = str.trim().split(SPLITTER_REGEX);
+            const parts = str.trim().split(SPLITTER_REGEX);
 
             _.forEach(parts, function (part) {
                 // try roman numbers
-                var matches = part.match(ROMAN_NUMBERS_REGEX);
+                let matches = part.match(ROMAN_NUMBERS_REGEX); // it isn't clear why there are for matches declaration
```
Comment (qooban) [inline src/Stations.js]: We want to identify stations and start/end points of the route, but we don't have strict rules at the moment for stations naming, so I try different regular expressions to find what is that point - 1. roman numbers, 2. european numbers, 3. start names, 4. end names. It's good idea to add comment above, then.

File: src/lodash.js [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
+const isEqual = require('lodash/isEqual');
+const isNull = require('lodash/isNull');
+const filter = require('lodash/filter');
+const find = require('lodash/find');
+const forEach = require('lodash/forEach');
+const map = require('lodash/map');
+const min = require('lodash/min');
+const sortBy = require('lodash/sortBy');
+
+export {isEmpty, isEqual, isNull, filter, find, forEach, map, min, sortBy};
```
Comment (qooban) [inline src/lodash.js]: The reason I have introduced such file was to reduce the output file size and I'm wondering how that works with your implementation using webpack&babel instead of browserify&uglifyjs. Could you check? 

I mean in all places where we import this files (`lodash.js`) we could normally import the whole `lodash` library. Will webpack&babel optimize it to include in bundle only those parts of lodash which are actually used?

File: src/routeVerifier.js [hunk 6]
```diff
@@ -1,84 +1,85 @@
...
+                .then(() => {
+                    const pathElevation = route.getPathElevation();
                     pathElevation.enrichData(routeLength);
 
-                    var isPathElevationGainValid = true;
+                    const isPathElevationGainValid = true;
                     controls.updateElevationGain(isPathElevationGainValid, pathElevation.gain);
 
-                    var isNormalRoute = routeLength >= 40 || pathElevation.gain > 500 && routeLength >= 30;
+                    const isNormalRoute = routeLength >= 40 || pathElevation.gain > 500 && routeLength >= 30;
```
Comment (qooban) [inline src/routeVerifier.js]: Let's convert this numbers into consts:
```
const NORMAL_ROUTE_MIN_LENGTH = 40; // kilometers
const SHORT_NORMAL_ROUTE_MIN_LENGTH = 30; // kilometers
const SHORT_NORMAL_ROUTE_MIN_ELEVATION_GAIN = 500; // meters
```

File: src/routeVerifier.js [hunk 7]
```diff
@@ -1,84 +1,85 @@
...
-                            var NORMAL_ROUTE_TYPE = 0;
-                            var INSPIRED_ROUTE_TYPE = 1;
+                        .then(parameters => {
+                            const ACCEPTED_ROUTE_LENGTH_DIFF = 1; //km
+                            const ACCEPTED_ELEVATION_GAIN_DIFF = 50; //m
+                            const NORMAL_ROUTE_TYPE = 0;
+                            const INSPIRED_ROUTE_TYPE = 1;
                             
-                            var isLengthConsistent = (routeLength - ACCEPTED_ROUTE_LENGTH_DIFF <= parameters.length &&
+                            const isLengthConsistent = (routeLength - ACCEPTED_ROUTE_LENGTH_DIFF <= parameters.length &&
```
Comment (qooban) [inline src/routeVerifier.js]: I'm wondering about code styling here... Not sure we should mix camel case and uppercased snake case. What do you think about taking code style improvements (check code against https://github.com/airbnb/javascript) and adding some kind of style checker (for example ESLint) after we finish with ES6? I would reate another issue for that.

File: package-lock.json [hunk 8]
```diff
@@ -1,3809 +0,0 @@
-{
```
Comment (qooban) [inline package-lock.json]: I think it should be part of the source code (see: https://stackoverflow.com/questions/44206782/do-i-commit-the-package-lock-json-file-created-by-npm-5), but I didn't go into details. If you think it shall not, let me know, we can discuss, I don't have detailed knowledge about that.

File: .gitignore (hunk 1 above)
Comment (turisap) [reply .gitignore]: fixed

File: package.json (hunk 2 above)
Comment (turisap) [reply package.json]: I tried to fix it and reinstall those packages, but this is how they are got installed from npm or babel website. Furthermore, when I run `npm view <package name> versions` it showed me that they are all beta. Is there something else I can do here?

File: src/lodash.js (hunk 5 above)
Comment (turisap) [reply src/lodash.js]: Surprisingly, `webpack` has not done such an optimization. I've replaced all imports from `lodash.js` to named imports from the library itself and then `npm run build:prod`. The resulting bundle was 130kB, while it was 86.9 kB while importing from file. So I guess we need to leave it as is.

File: src/routeVerifier.js (hunk 7 above)
Comment (turisap) [reply src/routeVerifier.js]: I like the idea to get familiar with ESLint would like to try to solve it on another issue)

File: src/routeVerifier.js (hunk 6 above)
Comment (turisap) [reply src/routeVerifier.js]: fixed

File: package-lock.json (hunk 8 above)
Comment (turisap) [reply package-lock.json]: Actually, I don't have deep knowledge on this as well.. But according to [npm docs](https://docs.npmjs.com/files/package-lock.json) `package-lock.json` should be checked in.

File: src/Stations.js (hunk 4 above)
Comment (turisap) [reply src/Stations.js]: I added one, but it's likely that I got the idea wrongly, please check it)

File: src/Route.js (hunk 3 above)
Comment (turisap) [reply src/Route.js]: Sorry, this is completely unclear for me)) I'd like to make these checks, but could you please give me more detailed steps for that?

File: src/lodash.js (hunk 5 above)
Comment (qooban) [reply src/lodash.js]: OK, thanks for checking that. Then yes, we'll leave it as is.

File: src/routeVerifier.js (hunk 7 above)
Comment (qooban) [reply src/routeVerifier.js]: Great, created https://github.com/edk-software/edk-route-verifier/issues/47. You can take it after we finish with this one.

File: src/Route.js (hunk 3 above)
Comment (qooban) [reply src/Route.js]: What I would like to achieve here is that we have proper handling for cases where KML file is lacking of some data which we are expecting to be there Our code should not brake because of that.

For example we have such KML file: https://github.com/edk-software/edk-route-verifier/blob/master/test/resources/1.kml

We can think of such problems:
1. What happen if we remove Placemark tag containing LineString tag? 
- in our naming convention => no path, 
- in KML naming convention => no LineString, 
- in GeoJSON naming convention => no LineString features
2. What happen if we remove all Placemark tags containing Point tags? 
- in our naming convention => no stations, 
- in KML naming convention => no Points, 
- in GeoJSON naming convention => no Point features

Few words about the background - the process of handling KML files at the moment:
1. Read KML file as XML document,
2. Convert KML to GeoJSON format (see here live demo of the lib which is used: https://mapbox.github.io/togeojson/)
3. After that all the operations we do are on GeoJSON format.

Comment (qooban) [reply src/Route.js]: @turisap Let me know if it is clearer now for you. That's the only thing we need to fix before I merge the code.

Comment (turisap) [reply src/Route.js]: @qooban, yeah, the idea is clear now, thanks for explanation)
I've changed slightly `constructor` in `Route.js` so that `station` and `path` would not initialize if checks don't pass in the following way:
```
if(this.isRouteVerifiable) {
            this.stations = new Stations(this.points, this.lineString);
            this.path = this.stations.isPathReversed() ? helpers.reverseLineString(this.lineString) : this.lineString;
        }
```
But now bundle fails the 13th test:
```
Running:  Positive test - circular path
 √ Element <body> was visible after 87 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 16 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 47 milliseconds.
 √ Element <div#pageReloadModal> was visible after 803 milliseconds.
 √ Element <div#pageReloadModal button.close> was visible after 78 milliseconds.
 √ Element <div#pageReloadModal> was not visible after 47 milliseconds.
-- Single path verification. Expecting path: single
 × Testing if element <div#singlePath span.info-box-icon> has css class: "bg-green".  - expected "has bg-green" but got: "info-box-icon bg-gray-light"
    at Page.assertSinglePath (D:\IT\Code\edk-route-verifier\test\pages\page.js:39:35)
    at Object.Positive test - circular path (D:\IT\Code\edk-route-verifier\test\tests\verification.js:99:14)


FAILED:  1 assertions failed and 6 passed (17.338s)
```

On the bundle without such modifications but when I deleted the `Placemark` tag which contained `LineString` tag tests failed here:
```
Running:  Positive test - short path
 √ Element <body> was visible after 78 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 31 milliseconds.
 × Timed out while waiting for element <button#verifyRoute div#loader> to be removed for 5000 milliseconds.  - expected "not found" but got: "found"
    at Page.verifyRoute (D:\IT\Code\edk-route-verifier\test\pages\page.js:23:22)
    at Object.Positive test - short path (D:\IT\Code\edk-route-verifier\test\tests\verification.js:10:14)


FAILED:  1 assertions failed and 2 passed (17.67s)

 _________________________________________________

 TEST FAILURE:  1 assertions failed, 2 passed. (22.102s)

 × verification

   - Positive test - short path (17.67s)
   Timed out while waiting for element <button#verifyRoute div#loader> to be removed for 5000 milliseconds.  - expected "not found" but got: "found"
```
while after deletion both `LineSting` and `Point` containing `Placemark` tags test failed here (which looks same with previous):

```
Running:  Positive test - short path
 √ Element <body> was visible after 89 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 32 milliseconds.
 × Timed out while waiting for element <button#verifyRoute div#loader> to be removed for 5000 milliseconds.  - expected "not found" but got: "found"
    at Page.verifyRoute (D:\IT\Code\edk-route-verifier\test\pages\page.js:23:22)
    at Object.Positive test - short path (D:\IT\Code\edk-route-verifier\test\tests\verification.js:10:14)


FAILED:  1 assertions failed and 2 passed (16.006s)

 _________________________________________________

 TEST FAILURE:  1 assertions failed, 2 passed. (21.779s)

 × verification

   - Positive test - short path (16.006s)
   Timed out while waiting fo
```

Comment (qooban) [reply src/Route.js]: I think you were on the right direction. Don't know why you get one of the tests failed.

I suggest applying such solution:
[0107-Route.js-update.txt](https://github.com/edk-software/edk-route-verifier/files/1812699/0107-Route.js-update.txt)

It will cover the issue I'm afraid of (KML file without LineString and/or Points). I'll add another issue to add some more tests for such cases (I have created some time ago placeholders for error scenarios and they are still empty ;-) ).

So - I have tested it locally and works fine - all the tests passed and when I created KML without LineString and/or Points and tried to use verifier on it, then it didn't crash and reported all checks as wrong in such way:
image
Could you add two tests to the verifier for such cases?
1. KML with some Points, but without LineString
2. KML with one LineString, but without Points
All output boxes shall be yellow. Do you need some guidance on how to create those tests? I'd like to have two Negative tests added to test/tests/verification.js

File: package.json [hunk 9]
```diff
@@ -14,7 +14,9 @@
     "git:push": "cross-var git push --tags origin HEAD:$npm_package_version",
     "deploy:clean": "git checkout master && cross-var git branch -D $npm_package_version",
     "deploy:current": "npm run bundle:test && npm test && npm run git:branch && npm run bundle:dist && npm run git:commit && cross-var git tag v$npm_package_version && npm run git:push",
-    "deploy:next": "npm run bundle:test && npm test && npm run git:branch && npm run bundle:dist && npm run git:commit && npm run version:patch && npm run git:push"
+    "deploy:next": "npm run bundle:test && npm test && npm run git:branch && npm run bundle:dist && npm run git:commit && npm run version:patch && npm run git:push",
+    "build:dev": "webpack --config webpack.config.js --mode development",
```
Comment (qooban) [inline package.json]: @turisap I have spotted two more things to do here:
1. Let's remove `bundle:test` and `bundle:dist` targets, we will use new webpack targets you created (you have already removed browserify and uglifyjs from package.json, so let's remove all usages). Please also update all other targets which used `bundle:...`.
2. We need also to update deploy mechanism. So I think we should create `build:dist` target which will do the same as `bundle:dist` = create bundle in normal and minified version in `dist` folder
Can you do that?

File: src/Route.js (hunk 3 above)
Comment (turisap) [reply src/Route.js]: I've fixed `constructor()` in the `Route` class, so now it looks as follows:
```
constructor (geoJson) {
        this.geoJson = geoJson;
        this.lineString = helpers.getLineString(this.geoJson);
        this.points = helpers.getPoints(this.geoJson);
        this.isRouteVerifiable = true;


        if(_.isEmpty(this.lineString)) {
            logger.error('No line string in route.');
            this.isRouteVerifiable = false;
            logger.error('No points in route.');
            this.isRouteVerifiable = false;
        }

        if(this.isRouteVerifiable) {
            this.stations = new Stations(this.points, this.lineString);
            this.path = this.stations.isPathReversed() ? helpers.reverseLineString(this.lineString) : this.lineString;
            this.numberOfPaths = helpers.getNumberOfFeatures('LineString', this.geoJson);
        }
    }
```
but there is no old check for empty `points` property like `_.isEmpty(points)`, is it a right thing?
> Could you add two tests to the verifier for such cases?

I think I could try it. I have only some experience with `Jest`/`enzyme`, not with `nightwatch` but I guess they should be similar somehow. I took a look on the test code and it looks like I would need explicit guidance to accomplish that :-)

Comment (qooban) [reply src/Route.js]: No, no, this check shall be there. It was not touched in the patch I shared. The code for the costructor shall be like that after applying the patch (`git patch ...`):
```
    constructor (geoJson) {
        this.geoJson = geoJson;
        this.lineString = helpers.getLineString(this.geoJson);
        this.points = helpers.getPoints(this.geoJson);
        this.isRouteVerifiable = true;

        if(_.isEmpty(this.lineString)) {
            logger.error('No line string in route.');
            this.isRouteVerifiable = false;
        }
        if(_.isEmpty(this.points)) {
            logger.error('No points in route.');
            this.isRouteVerifiable = false;
        }
        if(this.isRouteVerifiable) {
            this.stations = new Stations(this.points, this.lineString);
            this.path = this.stations.isPathReversed() ? helpers.reverseLineString(this.lineString) : this.lineString;
            this.numberOfPaths = helpers.getNumberOfFeatures('LineString', this.geoJson);
        }
    }
```

File: package.json (hunk 9 above)
Comment (turisap) [reply package.json]: > Let's remove bundle:test and bundle:dist targets, we will use new webpack targets you created (you have already removed browserify and uglifyjs from package.json, so let's remove all usages). 

Done

> Please also update all other targets which used bundle:....

Updated, take a look

> We need also to update deploy mechanism. So I think we should create build:dist target which will do the same as bundle:dist = create bundle in normal and minified version in dist folder

There is `build:dist` target now which creates minified bundle with source maps in `dist/prod` and non-minified bundle in `dist/dev` directories.

File: src/Route.js (hunk 3 above)
Comment (qooban) [reply src/Route.js]: In case of tests it should be quick. You would just need to:
1. Create KML file with no LineString (base on the existing one)
2. Add another test in `test/tests/verification.js`, sth like:
```
	 'Negative test - no path in KML file': function (client) {
        client.page.page()
            .navigate(<YOUR_KML_FILE_NAME_WITHOUT_EXTENSION>)
            .verifyRoute()
            .assertSinglePath(false)
            .assertNumberOfStations(false)
            .assertStationsOrder(false)
            .assertStationsOnPath(false)
            .assertDataConsistency(false);
    },
```
The same for the KML file without Points.

File: .gitignore [hunk 10]
```diff
@@ -6,9 +6,13 @@ node_modules/
...
+/server/static/js/edk-route-verifier.js.map
 /server/static/js/edk-route-verifier.js
 /test/screenshots/
 /test/reports/
 
 # Debug logs
 selenium-debug.log
+
+# Dist directory
+dist/
```
Comment (qooban) [inline .gitignore]: I'm not sure here, but let's leave it for now.

To show you broader perspective - I have taken similar approach as `jQuery` has - does not include bundles in dist folder in `master` branch, but includes it in tags. See:
- `master` - https://github.com/jquery/jquery/tree/master/dist
- `3.3.1` tag - https://github.com/jquery/jquery/tree/3.3.1/dist

I am using fetching `edk-route-verifier.js` or `edk-route-verifier.min.js` in `edk-software/rejony-app` using such base: https://cdn.jsdelivr.net/gh/edk-software/edk-route-verifier@1.0.1/dist/ .

File: webpack.config-dist.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+/**
```
Comment (qooban) [inline webpack.config-dist.js]: It's very similar to original configuration. 
If only difference is in output folder can't we parametrize it in `webpack.config.js` and not duplicate the configuration? What do you think?

File: package.json [hunk 12]
```diff
@@ -4,19 +4,18 @@
...
-    "deploy:current": "npm run bundle:test && npm test && npm run git:branch && npm run bundle:dist && npm run git:commit && cross-var git tag v$npm_package_version && npm run git:push",
-    "deploy:next": "npm run bundle:test && npm test && npm run git:branch && npm run bundle:dist && npm run git:commit && npm run version:patch && npm run git:push",
-    "build:dev": "webpack --config webpack.config.js --mode development",
-    "build:prod": "webpack --config webpack.config.js --mode production"
+    "deploy:current": "npm run build:prod && npm test && npm run git:branch && npm run build:dist && npm run git:commit && cross-var git tag v$npm_package_version && npm run git:push",
+    "deploy:next": "npm run build:prod && npm test && npm run git:branch && npm run build:dist && npm run git:commit && npm run version:patch && npm run git:push",
+    "build:dev" : "webpack --config webpack.config.js --mode development",
+    "build:prod": "webpack --config webpack.config.js --mode production",
+    "build:dist": "concurrently  \"webpack --config webpack.config-dist.js --mode production\" \"webpack --config webpack.config-dist.js --mode development\""
```
Comment (qooban) [inline package.json]: We should add info about all the stuff which needs to be installed globally (like webpack, concurrently) in `README.md` file. Could you update that file as well?

File: src/Route.js (hunk 3 above)
Comment (turisap) [reply src/Route.js]: Oh, sorry, I have never heard anything about patches in git)) I've made quick search and tried to run `git am <path to patch>` or `git apply <path to patch>` but this gave me only errors, so I just copied what you posted. It would be prefect to know what I was supposed to do with the patch file :-)

As for tests, I added both, created a new production build and tested it (I created two additional `.kml` files based on `1kml`: one with no `<LineString>`, and another without `<Point>` tags. And it passed all the tests!!! Actually, I'm not getting the idea behind it, but the only difference with your and my implementation is that you placed `this.numberOfPath` property in `if(this.isRouteVerifiable)` check in the class `constructor()`. Is it because of this?

Comment (qooban) [reply src/Route.js]: >It would be prefect to know what I was supposed to do with the patch file :-)

`git apply <path_to_patch>` in root folder of repo shall do the trick ;-)

>As for tests, I added both, created a new production build and tested it (I created two additional .kml files based on 1kml: one with no <LineString>, and another without <Point> tags. And it passed all the tests!!!

Great! Have you added tests for both cases? I see only for no-points case.


>I'm not getting the idea behind it, but the only difference with your and my implementation is that you placed this.numberOfPath property in if(this.isRouteVerifiable) check in the class constructor(). Is it because of this?

Actually I don't now, looks strange. Let's leave it.

File: .gitignore (hunk 10 above)
Comment (turisap) [reply .gitignore]: Thanks for explanation, actually I find it fun to work on `edk-route-verifier` because you are always meet my questions with detailed explanations which in turn helps me to get the job done and gain wanted experience, so I'm happy to participate while it comes with new knowledge :-)

Getting back to your comment, as far as I understood, I need to create a brunch named as a release number plus `tag`  , run `npm build: dist` and push that build into the respective brunch into the `upstream`. Is it right?

File: package.json (hunk 12 above)
Comment (turisap) [reply package.json]: Yes, sure, I've updated it. The one things is that `webpack` and `concurrently` are both in `package.json` deps list, so I'm not sure whether or not there is a need to install it globally (I added it for `webpack`, but `concurrently` should be fetched during `npm install`).

File: src/Route.js (hunk 3 above)
Comment (turisap) [reply src/Route.js]: > git apply <path_to_patch> in root folder of repo shall do the trick ;-)

Where should I place the patch file itself, also in the root?

> Great! Have you added tests for both cases? I see only for no-points case.

Sorry, I did `git reset` and missed that. I've already pushed it.

File: package.json (hunk 12 above)
Comment (qooban) [reply package.json]: OK, you're right until we run these commands directly from command line it is not necessary.

File: webpack.config-dist.js (hunk 11 above)
Comment (turisap) [reply webpack.config-dist.js]: It's a good idea actually and I've done that, but now test failing again:
```
Running:  Positive test - short path
 √ Element <body> was visible after 59 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 47 milliseconds.
 √ Element <button#verifyRoute div#loader> was not present after 16 milliseconds.
 × Timed out while waiting for element <div#pageReloadModal> to be visible for 5000 milliseconds.  - expected "visible" but got: "not visible"
    at Page.closePageReloadModal (D:\IT\Code\edk-route-verifier\test\pages\page.js:29:22)
    at Object.Positive test - short path (D:\IT\Code\edk-route-verifier\test\tests\verification.js:11:14)
```
I will check it later, it seems strange to me.

Comment (qooban) [reply webpack.config-dist.js]: OK, I would suspect problems with bundle file. When you create bundle file for distribution (`dist` folder), you won't have it placed in test server static js folder. We do want to use dist folder only for bundle deployment. For testing only test server folder will be used.

File: .gitignore (hunk 10 above)
Comment (qooban) [reply .gitignore]: I'm happy also :-)

>Getting back to your comment, as far as I understood, I need to create a brunch named as a release number plus tag , run npm build: dist and push that build into the respective brunch into the upstream. Is it right?

Roughly it is described here: https://github.com/edk-software/edk-route-verifier/blob/master/README.md#deployment
Let's leave it for now. I'll deploy after we finish with ES6 update and maybe I'll update documentaiton of that process.

File: webpack.config-dist.js (hunk 11 above)
Comment (turisap) [reply webpack.config-dist.js]: Sooo, this was a tricky one..

The culprit was filename and source maps. When I run `npm build:dev` it created `edk-router-verifer.min.js` and tests appeared  didn't work properly on this bundle, while after changing its name to `edk-route-verifier.js` it worked well. Next, when it came to the production build, after running `npm run build:prod` there were two files: `edk-router-verifier.js` and `edk-router-verifier.map.js`. So, when I  run `npm test` with two files it failed, but all the tests run successfully after deleting the latter! Furthermore, production and dev builds failed on different test cases which is also strange. 

I don't know what to change in test files, so I made all bundles regardless of mode named as `edk-route-verifier.js` and run test only after deleting source maps in the case of production build. And it passes.

**Comment by turisap** [comment]
@qooban, cool! I'm happy to see it merged!
